sturdy/constants.py

```diff
@@ -39,7 +39,8 @@
 SCORING_WINDOW = 300 # scoring window
 
 TOTAL_ALLOC_THRESHOLD = 0.98
-SIMILARITY_THRESHOLD = 0.01  # similarity threshold for plagiarism checking
+ALLOCATION_SIMILARITY_THRESHOLD = 1e-18  # similarity threshold for plagiarism checking
+APY_SIMILARITY_THRESHOLD = 1e-16
```

Comment:
why so smol 🤏?

Reply:
The issue with  high thresholds is that we are considering dissimilar allocations as identic. This creates a situation where only the fastest miners score well, and slower miners score poorly even if they have a higher APY. 

You can imagine a sphere in the plan of (pool_1, pool_2) allocations. The higher the threshold, the larger the sphere is. With a high threshold, the sphere will include many miners, and within the sphere, the latency will matter more than APY.

This happens because the miners are receiving assets and are sending back allocation with `18-21 decimals` represented as `int` then, the distance is calculated with enough precision, but the threshold is set way too high compared to the values of the distances between points.

I set the values to match  the lower end of the precision of the assets (1 wei) and the APY we see in practice.
By the way, I have set the thresholds such that there are few digits left depending on the assets sent. Miners with allocations 1000 or 100 wei apart will be considered suspiciously similar; currently, we consider cheating if the allocations are multiple eths apart.

Taking full precision is not perfect, but it aligns with the goal of the subnets and an increase in APY. 

### High Threshold

Only the fastest miners score well; the best miner scores poorly despite having the highest APY. This is so exacerbated on the subnet that you can increase your incentive by decreasing your APY.

```python
apys_and_allocations = {
    "0": {
        "allocations": {"pool_1": 495000000000000000000, "pool_2": 505000000000000000000}, # precision 20
        "alloc": np.array([gmpy2.mpz(495000000000000000000), gmpy2.mpz(505000000000000000000)], dtype=object),
        "apy": 0, # int(0.04e18), pecision 16
    },
    "1": {
        "allocations": {"pool_1": 495000000000000000000, "pool_2": 505000000000000000000}, # precision 20
        "alloc": np.array([gmpy2.mpz(495000000000000000000), gmpy2.mpz(505000000000000000000)], dtype=object),
        "apy": 40000000000000000, # int(0.04e18), pecision 16
    },
    "2": {
        "allocations": {"pool_1": 500000000000000000000, "pool_2": 500000000000000000000}, # precision 20
        "alloc": np.array([gmpy2.mpz(500000000000000000000), gmpy2.mpz(500000000000000000000)], dtype=object),
        "apy": 50000000000000000, # int(0.05e18), precision 16
    },
    "3": {
        "allocations": {"pool_1": 505000000000000000000, "pool_2": 495000000000000000000}, # precision 20
        "alloc": np.array([gmpy2.mpz(505000000000000000000), gmpy2.mpz(495000000000000000000)], dtype=object),
        "apy": 60000000000000000, # int(0.05e18), precision 16  
    },
}

axon_times = {"0": 45, "1": 1.0, "2": 2.0, "3": 3.0}
assets_and_pools = {
            "pools": {
                "pool_1": {"reserve_size": 100e18},
                "pool_2": {"reserve_size": 1000e18},
            },
            "total_assets": gmpy2.mpz(1000e18), # precision 21
        }
assets_and_pools = {
            "pools": {
                "pool_1": {"reserve_size": 100e18},
                "pool_2": {"reserve_size": 1000e18},
            },
            "total_assets": gmpy2.mpz(1000e18), # precision 21
        }


similarity_matrix = get_similarity_matrix(apys_and_allocations=apys_and_allocations,assets_and_pools=assets_and_pools)
penalties = calculate_penalties(similarity_matrix=similarity_matrix, axon_times=axon_times, similarity_threshold=SIMILARITY_THRESHOLD)
rewards_apy = dynamic_normalize_zscore(apys_and_allocations=apys_and_allocations)
adjusted_reward = calculate_rewards_with_adjusted_penalties(['0', '1', '2', '3'], rewards_apy, penalties)

f'Distance between miner 1 and 2: {get_distance(apys_and_allocations["1"]["alloc"], apys_and_allocations["2"]["alloc"], assets_and_pools["total_assets"])}'
f'Miners 1 and 2 allocations to close (threshold={SIMILARITY_THRESHOLD}): {get_distance(apys_and_allocations["1"]["alloc"], apys_and_allocations["2"]["alloc"], assets_and_pools["total_assets"]) < SIMILARITY_THRESHOLD}'
f'Distance between miner 1 and 3: {get_distance(apys_and_allocations["1"]["alloc"], apys_and_allocations["3"]["alloc"], assets_and_pools["total_assets"])}'
f'Miners 1 and 3 allocations to close (threshold={SIMILARITY_THRESHOLD}): {get_distance(apys_and_allocations["1"]["alloc"], apys_and_allocations["3"]["alloc"], assets_and_pools["total_assets"]) < SIMILARITY_THRESHOLD}'
f'Distance between miner 2 and 3: {get_distance(apys_and_allocations["2"]["alloc"], apys_and_allocations["3"]["alloc"], assets_and_pools["total_assets"])}'
f'Miners 2 and 3 allocations to close (threshold={SIMILARITY_THRESHOLD}): {get_distance(apys_and_allocations["2"]["alloc"], apys_and_allocations["3"]["alloc"], assets_and_pools["total_assets"]) < SIMILARITY_THRESHOLD}'

```

```bash
Similarity Matrix:

{'0': {'1': mpfr('0.0'),
  '2': mpfr('0.0050000000000000001'),
  '3': mpfr('0.01')},
 '1': {'0': mpfr('0.0'),
  '2': mpfr('0.0050000000000000001'),
  '3': mpfr('0.01')},
 '2': {'0': mpfr('0.0050000000000000001'),
  '1': mpfr('0.0050000000000000001'),
  '3': mpfr('0.0050000000000000001')},
 '3': {'0': mpfr('0.01'),
  '1': mpfr('0.01'),
  '2': mpfr('0.0050000000000000001')}}

Penalties:

{'0': 3, '1': 0, '2': 1, '3': 2}

Scores:

tensor([0.0000, 0.2500, 1.0000])

Reward:

tensor([0.0000, 0.4444, 0.6944, 1.0000])

Adjusted Reward:  # The order is messed up

tensor([0.0000, 0.4444, 0.4630, 0.3333])

'Distance between miner 1 and 2: 0.005000'
'Miners 1 and 2 allocations to close (threshold=0.1): True'
'Distance between miner 1 and 3: 0.010000'
'Miners 1 and 3 allocations to close (threshold=0.1): True'
'Distance between miner 2 and 3: 0.005000'
'Miners 2 and 3 allocations to close (threshold=0.1): True'

```

### Low Threshold

You can recover the proper order even for "smaller"  but not negligible differences 

```python
apys_and_allocations = {
    "0": {
        "allocations": {"pool_1": 495000000000000000000, "pool_2": 505000000000000000000}, # precision 20
        "alloc": np.array([gmpy2.mpz(495000000000000000000), gmpy2.mpz(505000000000000000000)], dtype=object),
        "apy": 0, # int(0.04e18), pecision 16
    },
    "1": {                         
        "allocations": {"pool_1": 499999999995000000000, "pool_2": 500000000005000000000}, # precision 20
        "alloc": np.array([gmpy2.mpz(499999999995000000000), gmpy2.mpz(500000000005000000000)], dtype=object),
        "apy": 40000000000000000, # int(0.04e18), pecision 16
    },
    "2": {
        "allocations": {"pool_1": 500000000000000000000, "pool_2": 500000000000000000000}, # precision 20
        "alloc": np.array([gmpy2.mpz(500000000000000000000), gmpy2.mpz(500000000000000000000)], dtype=object),
        "apy": 50000000000000000, # int(0.05e18), precision 16
    },
    "3": {
        "allocations": {"pool_1": 500000000005000000000, "pool_2": 499999999995000000000}, # precision 20
        "alloc": np.array([gmpy2.mpz(500000000005000000000), gmpy2.mpz(499999999995000000000)], dtype=object),
        "apy": 60000000000000000, # int(0.05e18), precision 16  
    },
}

axon_times = {"0": 45, "1": 1.0, "2": 2.0, "3": 3.0}
assets_and_pools = {
            "pools": {
                "pool_1": {"reserve_size": 100e18},
                "pool_2": {"reserve_size": 1000e18},
            },
            "total_assets": gmpy2.mpz(1000e18), # precision 21
        }


similarity_matrix = get_similarity_matrix(apys_and_allocations=apys_and_allocations,assets_and_pools=assets_and_pools)
penalties = calculate_penalties(similarity_matrix=similarity_matrix, axon_times=axon_times, similarity_threshold=SIMILARITY_THRESHOLD)
rewards_apy = dynamic_normalize_zscore(apys_and_allocations=apys_and_allocations)
adjusted_reward = calculate_rewards_with_adjusted_penalties([''0,'1', '2', '3'], rewards_apy, penalties)

f'Distance between miner 1 and 2: {get_distance(apys_and_allocations["1"]["alloc"], apys_and_allocations["2"]["alloc"], assets_and_pools["total_assets"])}'
f'Miners 1 and 2 allocations to close (threshold={1e-18}): {get_distance(apys_and_allocations["1"]["alloc"], apys_and_allocations["2"]["alloc"], assets_and_pools["total_assets"]) < 1e-18}'
f'Distance between miner 1 and 3: {get_distance(apys_and_allocations["1"]["alloc"], apys_and_allocations["3"]["alloc"], assets_and_pools["total_assets"])}'
f'Miners 1 and 3 allocations to close (threshold={1e-18}): {get_distance(apys_and_allocations["1"]["alloc"], apys_and_allocations["3"]["alloc"], assets_and_pools["total_assets"]) < 1e-18}'
f'Distance between miner 2 and 3: {get_distance(apys_and_allocations["2"]["alloc"], apys_and_allocations["3"]["alloc"], assets_and_pools["total_assets"])}'
f'Miners 2 and 3 allocations to close (threshold={1e-18): {get_distance(apys_and_allocations["2"]["alloc"], apys_and_allocations["3"]["alloc"], assets_and_pools["total_assets"]) <1e-18}'

```

```bash
Similarity Matrix:

{'0': {'1': mpfr('0.0049999999950000006'),
  '2': mpfr('0.0050000000000000001'),
  '3': mpfr('0.0050000000050000005')},
 '1': {'0': mpfr('0.0049999999950000006'),
  '2': mpfr('5.0000000000000005e-12'),
  '3': mpfr('1.0000000000000001e-11')},
 '2': {'0': mpfr('0.0050000000000000001'),
  '1': mpfr('5.0000000000000005e-12'),
  '3': mpfr('5.0000000000000005e-12')},
 '3': {'0': mpfr('0.0050000000050000005'),
  '1': mpfr('1.0000000000000001e-11'),
  '2': mpfr('5.0000000000000005e-12')}}

Penalties:

{'0': 0, '1': 0, '2': 0, '3': 0}

Scores:

tensor([0.0000, 0.4444, 0.6944, 1.0000])

Reward:

tensor([0.0000, 0.4444, 0.6944, 1.0000])

Adjusted Reward:  # The order is fixed

tensor([0.0000, 0.4444, 0.6944, 1.0000])

'Distance between miner 1 and 2: 0.005000'
'Miners 1 and 2 allocations to close (threshold=1e-18): False'
'Distance between miner 1 and 3: 0.010000'
'Miners 1 and 3 allocations to close (threshold=1e18): False'
'Distance between miner 2 and 3: 0.005000'
'Miners 2 and 3 allocations to close (threshold=1e-18): False'

```